### PR TITLE
merge #12196, #13697: Add scantxoutset RPC method

### PR DESCRIPTION
**Pull request by kittywhiskers**
# Overview

~~**This pull request presupposes the merger of https://github.com/dashpay/dash/pull/4186 into `develop`.** Only the last two commits are meant to be reviewed.~~ This pull request introduces the `scantxoutset` RPC command without segregated witness-specific logic. ~~To actually compare `psbt` with `scantxout`, go [here](https://github.com/kittywhiskers/dash/compare/psbt...kittywhiskers:scantxout)~~

## Motivation

During initial preparation for https://github.com/bitcoin/bitcoin/pull/18242 ("Add BIP324 encrypted p2p transport de-/serializer" by jonasschnelli) and the backporting of https://github.com/bitcoin/bitcoin/commit/78c312c983255e15fc274de2368a2ec13ce81cbf, a difference in class and function names were noticed and that in turn lead to a series of pull requests the lead us to this one.

The motivation is to reduce future merge conflicts when backporting pull requests from Bitcoin Core.

## Things To Do

~~* Fix `test/descriptor_tests.cpp` / possibly readapt them for Dash~~
```
test/descriptor_tests.cpp:53: error: in "descriptor_tests/descriptor_test": check parse_priv has failed
test/descriptor_tests.cpp:57: error: in "descriptor_tests/descriptor_test": check keys_priv.keys.size() has failed
unknown location:0: fatal error: in "descriptor_tests/descriptor_test": memory access violation at address: 0x00000000: no mapping at fault address
test/descriptor_tests.cpp:58: last checkpoint
```

## Contents

#12196, #13697

# Disclosures

* **This is a work in progress and needs external review**. Dash-specific changes have not been tested, only compilation has been ensured. Running the client on a testnet _may be_ necessary.
* ~~Tests fail. (see "Things To Do") Only Python ones do but at least we know how to get around it, see comment below~~

### CODE REVIEW

**Comment by kittywhiskers** [comment]
Rebased and updated, Python tests pending (inherited from PSBT)

**Comment by kittywhiskers** [comment]
Rebased on top of develop, post PSBT merge!

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -2273,6 +2278,217 @@ static UniValue savemempool(const JSONRPCRequest& request)
...
+{
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
+        throw std::runtime_error(
+            "scantxoutset <action> ( <scanobjects> )\n"
+            "\nEXPERIMENTAL warning: this call may be removed or changed in future releases.\n"
+            "\nScans the unspent transaction output set for entries that match certain output descriptors.\n"
+            "Examples of output descriptors are:\n"
+            "    addr(<address>)                      Outputs whose scriptPubKey corresponds to the specified address (does not include P2PK)\n"
+            "    raw(<hex script>)                    Outputs whose scriptPubKey equals the specified hex scripts\n"
+            "    combo(<pubkey>)                      P2PK, P2PKH, P2WPKH, and P2SH-P2WPKH outputs for the given pubkey\n"
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/rpc/blockchain.cpp:2351]: https://github.com/sipa/bitcoin/blob/f6b7fc349ccf9cfbeb7e91e19c20e2a2fcc9026f/src/script/descriptor.cpp#L355-L360

[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/rpc/blockchain.cpp]: Sure, we won't support segwit part of it but P2PK and P2PKH https://github.com/sipa/bitcoin/blob/f6b7fc349ccf9cfbeb7e91e19c20e2a2fcc9026f/src/script/descriptor.cpp#L349-L352 are legacy scripts we do support so why dropping it?

[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/rpc/blockchain.cpp]: Fair, I was interesting in maintaining 1:1 functionality so combo didn't fit that so I yanked it out

File: src/test/descriptor_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+    // Check for invalid nesting of structures
+    CheckUnparsable("sh(XJvEUEcFWCHCyruc8ZX5exPZaGe4UR7gC5FHrhwPnQGDs1uWCsT2)", "sh(03a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd)"); // P2SH needs a script, not a key
+    CheckUnparsable("sh(sh(pk(XJvEUEcFWCHCyruc8ZX5exPZaGe4UR7gC5FHrhwPnQGDs1uWCsT2)))", "sh(sh(pk(03a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd)))"); // Cannot embed P2SH inside P2SH
```
Comment (UdjinM6) [inline src/test/descriptor_tests.cpp:137]: same for
```
CheckUnparsable("sh(combo(L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1))", "sh(combo(03a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd))"); // Old must be top level
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Comment by kittywhiskers** [comment]
```
# L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1
**f58c277bb348b0ab476b16c1635bc8eb711daf5cc7965aae11d020f6fdfde8f7__--------

SHA256d: b41738f51164e711c444441c5050368b0d1ceb2ef138f23e08e37df0ad52e29b
Final: cce3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b85501b41738f5
Generated: XJvEUEcFWCHCyruc8ZX5exPZaGe4UR7gC5FHrhwPnQGDs1uWCsT2

# 5KYZdUEo39z3FPrtuX2QbbwGnNP5zTd7yyr2SC1j299sBCnWjss
**f58c277bb348b0ab476b16c1635bc8eb711daf5cc7965aae11d020f6fdfde8f7__--------

SHA256d: caf1acd1bebe5fac7e7e0d816c2dcbafa23fee3adc72ba513c741ef40f429b5f
Final: cce3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855caf1acd1
Generated: 7sH936MDoVPFVk2VoaCM5yW8P3BfPyffnZECyaHrZwfLgWpS13e

Running 380 test cases...
test/descriptor_tests.cpp:91: error: in "descriptor_tests/descriptor_test": check spks.size() == ref.size() has failed [2 != 4]
test/descriptor_tests.cpp:91: error: in "descriptor_tests/descriptor_test": check spks.size() == ref.size() has failed [2 != 4]
test/descriptor_tests.cpp:91: error: in "descriptor_tests/descriptor_test": check spks.size() == ref.size() has failed [2 != 4]
test/descriptor_tests.cpp:91: error: in "descriptor_tests/descriptor_test": check spks.size() == ref.size() has failed [2 != 4]
test/descriptor_tests.cpp:91: error: in "descriptor_tests/descriptor_test": check spks.size() == ref.size() has failed [2 != 4]
test/descriptor_tests.cpp:91: error: in "descriptor_tests/descriptor_test": check spks.size() == ref.size() has failed [2 != 4]
```

@UdjinM6

**Comment by UdjinM6** [comment]
pls see 7dacf7a4e9

**Review by UdjinM6 — APPROVED** [review]
LGTM, utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Pull request merged**